lib/ecto/enum.ex

```diff
@@ -46,17 +50,25 @@ defmodule Ecto.Enum do
   def init(opts) do
     values = Keyword.get(opts, :values, nil)
 
-    unless is_list(values) and Enum.all?(values, &is_atom/1) do
+    unless (is_list(values) and
+              Enum.all?(values, &is_atom/1)) or
+             (Keyword.keyword?(values) and Enum.all?(Keyword.values(values), &is_integer/1)) do
       raise ArgumentError, """
       Ecto.Enum types must have a values option specified as a list of atoms. For example:
 
           field :my_field, Ecto.Enum, values: [:foo, :bar]
       """
     end
 
-    on_load = Map.new(values, &{Atom.to_string(&1), &1})
-    on_dump = Map.new(values, &{&1, Atom.to_string(&1)})
-    %{on_load: on_load, on_dump: on_dump, values: values}
+    if Enum.all?(values, &is_atom/1) do
+      on_load = Map.new(values, &{Atom.to_string(&1), &1})
+      on_dump = Map.new(values, &{&1, Atom.to_string(&1)})
+      %{on_load: on_load, on_dump: on_dump, values: values}
+    else
+      on_load = Map.new(values, fn {key, val} -> {val, key} end)
+      on_dump = Enum.into(values, %{})
+      %{on_load: on_load, on_dump: on_dump, values: Keyword.keys(values)}
```

Comment:
This is good! But now that I think about it.... maybe we should stop storing `values` and use `Map.keys(on_dump)` instead? :D

Reply:
That is worth investigating. I just tried changing it to use ```Map.keys(on_dump)``` instead but had failing test, because the ordering of the values (sometimes) changed?

Reply:
Ah, converting to a map will change the order indeed. Ok, let's keep it then. :D